src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py

```diff
@@ -102,6 +102,7 @@ async def parse_python_dependencies(
                 "ASSETS_MIN_SLASHES": str(request.assets_min_slashes),
             },
             level=LogLevel.DEBUG,
+            docker_image=None,  # TODO(#7735): is this right?
```

Comment:
@stuhood this is a good example of how I don't understand how to reason about when Docker images can be used (or remote execution!), vs requiring local execution. Does it make sense to parse Python imports in Docker? That adds overhead. But it does make sense with RE due to parallelization!

I'm wondering if rather than `Process` callers setting `docker_image`, the rule graph should automatically set it via `ResolvedEnvironmentTarget`?

Reply:
> Does it make sense to parse Python imports in Docker?

It does if the host machine doesn't have Python, right? (Strange, but for argument's sake...)

Reply:
Also, @stuhood if we do add `NormalizedProcess`, I'm trying to think about how to model this in PyO3. Probably define that with Rust, right?

Reply:
Yea, this is a very fundamental question that we're going to want to experiment with the answer to.

As it stands with the `EnvironmentName` in most (but not all: only portions which actually consume it indirectly via a `Platform`/`Process`/option, etc) portions of the graph, there is a high likelihood of redundant work unless some portions of the graph are pinned back to the current environment.

----

My initial feeling (with not much evidence) is that we're going to want to always pin all `Dependencies` computation (`TransitiveTargets`/`CoarsenedTargets`/`DependenciesRequest`/etc) to the `__local__` environment.

The reasoning for that is that we already know that it is working for our existing users: even if they deploy to production using a different platform, their users currently have enough set up in their local environments (be it Linux or macOS) to allow for dependencies to be computed... and probably also for `check`/`lint`/etc to run.

While we could definitely enable new use cases by running the entire graph in the docker environment (i.e. even the "root" environment computed in #16786 would be a docker environment), those are the use cases which users currently satisfy by running the entire Pants process inside docker. So that would be a nice to have, but less important than just ensuring that your _**runtime**_ environment is a particular platform.

So yea: we initially care primarily about controlling the _runtime_ environment. Given that, dependency computation seems like it can/should be pinned to `__local__`, to allow it to be shared across all environments. But having said that: I _think_ that avoiding re-computing dependencies in a bunch of environments is primarily a performance concern rather than a correctness one (fairly unsure here), so I think that opening an issue to audit the impact in situations where we're running tests across multiple environments might be fine (rather than worrying about it in this PR).

Reply:
Thank you! Hmmm so I think correctness wise, it's wrong to be manually setting `docker_image` on each `Process` like this PR does. Why? If `EnvironmentName` points to a docker env, then it would be wrong to run with `docker_image=None`, which implies running locally. You would run locally, but with the Environment for the Docker image. Does that sound right?

That would imply we need to make this automated via something like `Process -> NormalizedProcess`, or in the past you've proposed that Rust intrinsics code grabs the value.